### ISSUE TITLE
SoftwareSerial bug - local variable, dangling pointer

### ISSUE THREAD

**Issue by heX16** [avaldebe/PMserial#19]
I look at [this code](https://github.com/avaldebe/PMserial/blob/98ccdffcfd50219e522fd0830552d082bae824e7/src/PMserial.h#L88) and see an error ([explain in cppreference](https://en.cppreference.com/w/cpp/language/reference#Dangling_references)).
The variable `serial` is local (in stack), its lifetime is limited to the area of the constructor. After exiting the constructor, the variable `serial` will be destroyed,  and the variable `uart` will contain [dangling pointer](https://en.wikipedia.org/wiki/Dangling_pointer). Copying its address should not prolong the lifetime.

I tried looking for some special case that affects the lifetime of a variable and which will not be "undefined behavior" ([0](https://en.cppreference.com/w/cpp/language/lifetime), [1](https://en.cppreference.com/w/cpp/language/reference#Dangling_references), [2](https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary), [3](https://stackoverflow.com/questions/22114751/returning-static-pointer-to-local-variable-from-function), [4](https://qna.habr.com/q/246744), [5](https://habr.com/ru/company/jugru/blog/469465/), [6](https://habr.com/ru/post/186790/) ). But I couldn't find anything suitable for this code...

If my fears are in vain, please give me a link to a description of this case in the standard or explain how it works.

Problem code:
```
class SerialPM
{
protected:
  Stream *uart;  // hardware/software serial
....

  SerialPM(PMS sensor, uint8_t rx, uint8_t tx) : pms(sensor)
  {
    SoftwareSerial serial(rx, tx);
    uart = &serial;
    hwSerial = false;
  }
```

**Comment by avaldebe** [comment]
I think you're, right. If I remember correctly, I tested this feature with actual HW when I implemented this feature, but I probably hit an undefined behavior of the uC tool-chain.

Will you be interested on submitting a patch/PR?

**Comment by heX16** [comment]
> Will you be interested on submitting a patch/PR?

Yes! I have an idea to declare "Serial" outside and pass it to pointer initialization (it will be completely analogous to "HardwareSerial"). And make the incoming pointer type "[Stream](https://github.com/arduino/ArduinoCore-avr/blob/6ec80154cd2ca52bce443afbbed8a1ad44d049cb/libraries/SoftwareSerial/src/SoftwareSerial.h#L50)". This will allow you to use any "Serial" and even their non-standard implementations.
I'll add code and correct examples of working with "SoftwareSerial". Work with "HardwareSerial" will not change.

**Comment by avaldebe** [comment]
> > Will you be interested on submitting a patch/PR?
> 
> Yes! I have an idea to declare "Serial" outside and pass it to pointer initialization (it will be completely analogous to "HardwareSerial"). And make the incoming pointer type "[Stream](https://github.com/arduino/ArduinoCore-avr/blob/6ec80154cd2ca52bce443afbbed8a1ad44d049cb/libraries/SoftwareSerial/src/SoftwareSerial.h#L50)". This will allow you to use any "Serial" and even their non-standard implementations.
> I'll add code and correct examples of working with "SoftwareSerial". Work with "HardwareSerial" will not change.

Awesome!

**Comment by avaldebe** [comment]
fixed on #20
